src/transforms/sample/transform.rs

```diff
@@ -69,13 +80,46 @@ impl FunctionTransform for Sample {
             })
             .map(|v| v.to_string_lossy());
 
+        // Fetch actual field value if group_by option is set.
+        let group_by_value = self
+            .group_by
+            .as_ref()
+            .and_then(|group_by| match &event {
+                Event::Log(event) => event
+                    .parse_path_and_get_value(group_by.as_str())
+                    .ok()
+                    .flatten(),
+                Event::Trace(event) => event
+                    .parse_path_and_get_value(group_by.as_str())
+                    .ok()
+                    .flatten(),
+                Event::Metric(_) => panic!("component can never receive metric events"),
+            })
+            .map(|v| v.to_string());
+
+        // Find the appropriate counter_key: If group_by option is passed, the counter key should
+        // be the value of the log attribute. If group_by option is not passed, then it should just
+        // fallback to the default bucket (i.e. have the same functionality as a general counter).
+        let counter_key: String = if let Some(group_by_value) = group_by_value {
+            group_by_value
+        } else {
+            default_group_name
+        };
+
+        let counter_value: u64 = match self.counter.entry(counter_key.clone()) {
+            Entry::Occupied(e) => *e.get(),
+            Entry::Vacant(e) => *e.insert(0),
+        };
```

Comment:
```suggestion
        let counter_value: u64 = match self.counter.entry(counter_key.clone()).or_default();
```

I think this would work?